// node-ntcore/dist/networktables/network-table-value.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClassCastError = void 0;
const network_table_instance_1 = require("./network-table-instance");
class ClassCastError extends Error {
}
exports.ClassCastError = ClassCastError;
class NetworkTableValue {
    constructor(type, value, time = Date.now()) {
        this._type = type;
        this._value = value;
        this._time = time;
    }
    // STATIC METHODS
    static makeBoolean(val, time = Date.now()) {
        return new NetworkTableValue(network_table_instance_1.NetworkTableType.BOOLEAN, { bool: val }, time);
    }
    static makeDouble(val, time = Date.now()) {
        return new NetworkTableValue(network_table_instance_1.NetworkTableType.DOUBLE, { double: val }, time);
    }
    static makeString(val, time = Date.now()) {
        return new NetworkTableValue(network_table_instance_1.NetworkTableType.STRING, { str: val }, time);
    }
    static makeRaw(val, time = Date.now()) {
        return new NetworkTableValue(network_table_instance_1.NetworkTableType.RAW, { raw: val }, time);
    }
    static makeBooleanArray(val, time = Date.now()) {
        return new NetworkTableValue(network_table_instance_1.NetworkTableType.BOOLEAN_ARRAY, { bool_array: val }, time);
    }
    static makeDoubleArray(val, time = Date.now()) {
        return new NetworkTableValue(network_table_instance_1.NetworkTableType.DOUBLE_ARRAY, { double_array: val }, time);
    }
    static makeStringArray(val, time = Date.now()) {
        return new NetworkTableValue(network_table_instance_1.NetworkTableType.STRING_ARRAY, { str_array: val }, time);
    }
    static makeUnassigned(time = Date.now()) {
        return new NetworkTableValue(network_table_instance_1.NetworkTableType.UNASSIGNED, {}, time);
    }
    getType() {
        return this._type;
    }
    getTime() {
        return this._time;
    }
    isValid() {
        return this._type !== network_table_instance_1.NetworkTableType.UNASSIGNED;
    }
    isBoolean() {
        return this._type === network_table_instance_1.NetworkTableType.BOOLEAN;
    }
    isBooleanArray() {
        return this._type === network_table_instance_1.NetworkTableType.BOOLEAN_ARRAY;
    }
    isDouble() {
        return this._type === network_table_instance_1.NetworkTableType.DOUBLE;
    }
    isDoubleArray() {
        return this._type === network_table_instance_1.NetworkTableType.DOUBLE_ARRAY;
    }
    isString() {
        return this._type === network_table_instance_1.NetworkTableType.STRING;
    }
    isStringArray() {
        return this._type === network_table_instance_1.NetworkTableType.STRING_ARRAY;
    }
    isRaw() {
        return this._type === network_table_instance_1.NetworkTableType.RAW;
    }
    isRPC() {
        return this._type === network_table_instance_1.NetworkTableType.RPC;
    }
    getBoolean() {
        if (this._type !== network_table_instance_1.NetworkTableType.BOOLEAN) {
            throw new ClassCastError("Cannot convert to BOOLEAN");
        }
        return this._value.bool;
    }
    getBooleanArray() {
        if (this._type !== network_table_instance_1.NetworkTableType.BOOLEAN_ARRAY) {
            throw new ClassCastError("Cannot convert to BOOLEAN_ARRAY");
        }
        return this._value.bool_array;
    }
    getDouble() {
        if (this._type !== network_table_instance_1.NetworkTableType.DOUBLE) {
            throw new ClassCastError("Cannot convert to DOUBLE");
        }
        return this._value.double;
    }
    getDoubleArray() {
        if (this._type !== network_table_instance_1.NetworkTableType.DOUBLE_ARRAY) {
            throw new ClassCastError("Cannot convert to DOUBLE_ARRAY");
        }
        return this._value.double_array;
    }
    getString() {
        if (this._type !== network_table_instance_1.NetworkTableType.STRING) {
            throw new ClassCastError("Cannot convert to STRING");
        }
        return this._value.str;
    }
    getStringArray() {
        if (this._type !== network_table_instance_1.NetworkTableType.STRING_ARRAY) {
            throw new ClassCastError("Cannot convert to STRING_ARRAY");
        }
        return this._value.str_array;
    }
    getRaw() {
        if (this._type !== network_table_instance_1.NetworkTableType.RAW) {
            throw new ClassCastError("Cannot convert to RAW");
        }
        return this._value.raw;
    }
}
exports.default = NetworkTableValue;
