// node-ntcore/dist/networktables/network-table-entry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NetworkTableEntryFlags = void 0;
var NetworkTableEntryFlags;
(function (NetworkTableEntryFlags) {
    NetworkTableEntryFlags[NetworkTableEntryFlags["UNASSIGNED"] = 0] = "UNASSIGNED";
    NetworkTableEntryFlags[NetworkTableEntryFlags["PERSISTENT"] = 1] = "PERSISTENT";
})(NetworkTableEntryFlags = exports.NetworkTableEntryFlags || (exports.NetworkTableEntryFlags = {}));
/**
 * NetworkTableEntry accessor
 *
 * A NetworkTableEntry provides a handle to a particular NT entry.
 * This NT entry could be real (exists on the server/network) or
 * pseudo (is pending transmission to the network)
 */
class NetworkTableEntry {
    constructor(inst, key, guid, funcs) {
        this._instance = inst;
        this._funcs = funcs;
        this._guid = guid;
        this._key = key;
    }
    getName() {
        return this._key;
    }
    getLastChange() {
        return this._funcs.getLastChange(this._key);
    }
    getType() {
        return this._funcs.getType(this._key);
    }
    getValue() {
        return this._funcs.getValue(this._key);
    }
    getBoolean(defaultVal) {
        return this._funcs.getBoolean(this._key, defaultVal);
    }
    getDouble(defaultVal) {
        return this._funcs.getDouble(this._key, defaultVal);
    }
    getString(defaultVal) {
        return this._funcs.getString(this._key, defaultVal);
    }
    getBooleanArray(defaultVal) {
        return this._funcs.getBooleanArray(this._key, defaultVal);
    }
    getDoubleArray(defaultVal) {
        return this._funcs.getDoubleArray(this._key, defaultVal);
    }
    getStringArray(defaultVal) {
        return this._funcs.getStringArray(this._key, defaultVal);
    }
    getRaw(defaultVal) {
        return this._funcs.getRaw(this._key, defaultVal);
    }
    setBoolean(val) {
        return this._funcs.setBoolean(this._key, val);
    }
    setDouble(val) {
        return this._funcs.setDouble(this._key, val);
    }
    setString(val) {
        return this._funcs.setString(this._key, val);
    }
    setBooleanArray(val) {
        return this._funcs.setBooleanArray(this._key, val);
    }
    setDoubleArray(val) {
        return this._funcs.setDoubleArray(this._key, val);
    }
    setStringArray(val) {
        return this._funcs.setStringArray(this._key, val);
    }
    setRaw(val) {
        return this._funcs.setRaw(this._key, val);
    }
    exists() {
        return this._funcs.exists(this._key);
    }
    delete() {
        return this._funcs.delete(this._key);
    }
    getFlags() {
        return this._funcs.getFlags(this._key);
    }
    setFlags(flags) {
        this._funcs.setFlags(this._key, this.getFlags() | flags);
    }
    clearFlags(flags) {
        this._funcs.setFlags(this._key, this.getFlags() & ~flags);
    }
    setPersistent() {
        this.setFlags(NetworkTableEntryFlags.PERSISTENT);
    }
    clearPersistent() {
        this.clearFlags(NetworkTableEntryFlags.PERSISTENT);
    }
    isPersistent() {
        return (this.getFlags() & NetworkTableEntryFlags.PERSISTENT) !== 0;
    }
}
exports.default = NetworkTableEntry;
