// node-ntcore/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogLevel = exports.NetworkTableEntryFlags = exports.NetworkTableEntry = exports.NetworkTable = exports.ConnectionState = exports.OperatingMode = exports.EntryListenerFlags = exports.NetworkTableInstance = void 0;
// `networktables` package
const network_table_instance_1 = __importStar(require("./networktables/network-table-instance"));
exports.NetworkTableInstance = network_table_instance_1.default;
Object.defineProperty(exports, "EntryListenerFlags", { enumerable: true, get: function () { return network_table_instance_1.EntryListenerFlags; } });
Object.defineProperty(exports, "OperatingMode", { enumerable: true, get: function () { return network_table_instance_1.OperatingMode; } });
Object.defineProperty(exports, "ConnectionState", { enumerable: true, get: function () { return network_table_instance_1.ConnectionState; } });
const network_table_1 = __importDefault(require("./networktables/network-table"));
exports.NetworkTable = network_table_1.default;
const network_table_entry_1 = __importStar(require("./networktables/network-table-entry"));
exports.NetworkTableEntry = network_table_entry_1.default;
Object.defineProperty(exports, "NetworkTableEntryFlags", { enumerable: true, get: function () { return network_table_entry_1.NetworkTableEntryFlags; } });
const log_util_1 = require("./utils/log-util");
Object.defineProperty(exports, "LogLevel", { enumerable: true, get: function () { return log_util_1.LogLevel; } });
