// node-ntcore/dist/protocol/nt-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NTEventUpdateSource = exports.NTEntryTypeMismatchError = exports.NTEntryNotFoundError = exports.NTProtocolVersionUnsupportedError = exports.NTConnectionState = void 0;
var NTConnectionState;
(function (NTConnectionState) {
    NTConnectionState[NTConnectionState["NTCONN_NOT_CONNECTED"] = 0] = "NTCONN_NOT_CONNECTED";
    NTConnectionState[NTConnectionState["NTCONN_CONNECTING"] = 1] = "NTCONN_CONNECTING";
    NTConnectionState[NTConnectionState["NTCONN_CONNECTED"] = 2] = "NTCONN_CONNECTED";
})(NTConnectionState = exports.NTConnectionState || (exports.NTConnectionState = {}));
class NTProtocolVersionUnsupportedError extends Error {
    constructor(serverSupportedVersion, msg) {
        super(msg);
        this._serverSupportedVersion = serverSupportedVersion;
    }
    get serverSupportedVersion() {
        return this._serverSupportedVersion;
    }
}
exports.NTProtocolVersionUnsupportedError = NTProtocolVersionUnsupportedError;
class NTEntryNotFoundError extends Error {
}
exports.NTEntryNotFoundError = NTEntryNotFoundError;
class NTEntryTypeMismatchError extends Error {
}
exports.NTEntryTypeMismatchError = NTEntryTypeMismatchError;
var NTEventUpdateSource;
(function (NTEventUpdateSource) {
    NTEventUpdateSource[NTEventUpdateSource["LOCAL"] = 1] = "LOCAL";
    NTEventUpdateSource[NTEventUpdateSource["REMOTE"] = 2] = "REMOTE";
})(NTEventUpdateSource = exports.NTEventUpdateSource || (exports.NTEventUpdateSource = {}));
